File: src/Hedge/Codec.fs
module Hedge.Codec

/// Generic runtime codec engine.
///
/// Uses Fable's TypeInfo reflection to auto-generate Thoth.Json
/// encoders and decoders for any record type, including wrapper
/// DU fields (PrimaryKey, RichContent, etc.).
///
/// Public API:
///   encode<'T> value  — encode a record to JsonValue
///   decode<'T> ()     — produce a Decoder<'T> for a record

open Fable.Core
open Thoth.Json

// ============================================================
// JS interop helpers
// ============================================================

/// Property access: obj[name]
[<Emit("$0[$1]")>]
let private getProp (obj: obj) (name: string) : obj = jsNative

/// Unwrap a single-case DU: value.fields[0]
[<Emit("$0.fields[0]")>]
let private unwrapDU (v: obj) : obj = jsNative

/// Null/undefined check (loose equality catches both)
[<Emit("($0 == null)")>]
let private isNull (v: obj) : bool = jsNative

/// Construct a single-case DU: new typeInfo.construct(arg)
/// Fable optimizes single-case DUs — constructor takes just the
/// inner value; the tag is hardcoded to 0 inside.
[<Emit("new ($0.construct)($1)")>]
let private constructDU (typeInfo: System.Type) (arg: obj) : obj = jsNative

/// Construct a record: new typeInfo.construct(...values)
[<Emit("new ($0.construct)(...$1)")>]
let private makeRecord (typeInfo: System.Type) (values: obj array) : obj = jsNative

// ============================================================
// TypeInfo accessors (same pattern as Schema.fs)
// ============================================================

[<Emit("$0.fullname")>]
let private tiFullname (t: System.Type) : string = jsNative

[<Emit("$0.generics || []")>]
let private tiGenerics (t: System.Type) : System.Type array = jsNative

[<Emit("$0.fields ? $0.fields() : []")>]
let private tiFields (t: System.Type) : (string * System.Type) array = jsNative

// ============================================================
// Helpers
// ============================================================

/// PascalCase → camelCase
let private camelCase (s: string) =
    if s.Length = 0 then s
    else string (System.Char.ToLowerInvariant s.[0]) + s.[1..]

/// Classify a wrapper DU type. Returns the base type name
/// ("string" or "int") if it's a known wrapper, None otherwise.
let private wrapperBaseType (fullname: string) (generics: System.Type array) : string option =
    if fullname.Contains("Interface.PrimaryKey") then
        if generics.Length > 0 && (tiFullname generics.[0]).Contains("Int32") then Some "int"
        else Some "string"
    elif fullname.Contains("Interface.MultiTenant") then Some "string"
    elif fullname.Contains("Interface.CreateTimestamp") then Some "int"
    elif fullname.Contains("Interface.UpdateTimestamp") then Some "int"
    elif fullname.Contains("Interface.SoftDelete") then Some "int"
    elif fullname.Contains("Interface.ForeignKey") then Some "string"
    elif fullname.Contains("Interface.RichContent") then Some "string"
    elif fullname.Contains("Interface.Link") then Some "string"
    else None

// ============================================================
// Encoder
// ============================================================

let rec encodeValue (typeInfo: System.Type) (value: obj) : JsonValue =
    let fn = tiFullname typeInfo
    let gs = tiGenerics typeInfo

    match wrapperBaseType fn gs with
    | Some "int" -> Encode.int (unbox (unwrapDU value))
    | Some _ -> Encode.string (unbox (unwrapDU value))
    | None ->
        if fn.Contains("String") then Encode.string (unbox value)
        elif fn.Contains("Int32") then Encode.int (unbox value)
        elif fn.Contains("Boolean") then Encode.bool (unbox value)
        elif fn.Contains("FSharpList") then
            let innerTi = if gs.Length > 0 then gs.[0] else typeInfo
            let items : obj list = unbox value
            items |> List.map (encodeValue innerTi) |> Encode.list
        else
            encodeRecord typeInfo value

and encodeRecord (typeInfo: System.Type) (value: obj) : JsonValue =
    let fs = tiFields typeInfo
    let pairs =
        fs |> Array.map (fun (name, fieldTi) ->
            let jsonName = camelCase name
            let fieldVal = getProp value name
            let fn = tiFullname fieldTi
            let gs = tiGenerics fieldTi

            if fn.Contains("FSharpOption") && gs.Length > 0 then
                let innerTi = gs.[0]
                if isNull fieldVal then
                    jsonName, Encode.nil
                else
                    jsonName, encodeValue innerTi fieldVal
            else
                jsonName, encodeValue fieldTi fieldVal
        )
    Encode.object (Array.toList pairs)

// ============================================================
// Decoder
// ============================================================

let rec decoderFor (typeInfo: System.Type) : Decoder<obj> =
    let fn = tiFullname typeInfo
    let gs = tiGenerics typeInfo

    match wrapperBaseType fn gs with
    | Some "int" ->
        Decode.int |> Decode.map (fun i -> constructDU typeInfo (box i))
    | Some _ ->
        Decode.string |> Decode.map (fun s -> constructDU typeInfo (box s))
    | None ->
        if fn.Contains("String") then Decode.string |> Decode.map box
        elif fn.Contains("Int32") then Decode.int |> Decode.map box
        elif fn.Contains("Boolean") then Decode.bool |> Decode.map box
        elif fn.Contains("FSharpList") then
            let innerTi = if gs.Length > 0 then gs.[0] else typeInfo
            Decode.list (decoderFor innerTi) |> Decode.map box
        else
            decodeRecordObj typeInfo

and decodeRecordObj (typeInfo: System.Type) : Decoder<obj> =
    Decode.object (fun get ->
        let fs = tiFields typeInfo
        let values =
            fs |> Array.map (fun (name, fieldTi) ->
                let jsonName = camelCase name
                decodeFieldValue get jsonName fieldTi
            )
        makeRecord typeInfo values
    )

and decodeFieldValue (get: Decode.IGetters) (jsonName: string) (fieldTi: System.Type) : obj =
    let fn = tiFullname fieldTi
    let gs = tiGenerics fieldTi

    if fn.Contains("FSharpOption") && gs.Length > 0 then
        let innerTi = gs.[0]
        let innerFn = tiFullname innerTi
        let innerGs = tiGenerics innerTi

        match wrapperBaseType innerFn innerGs with
        | Some "int" ->
            let optVal : int option = get.Optional.Field jsonName Decode.int
            box (optVal |> Option.map (fun i -> constructDU innerTi (box i)))
        | Some _ ->
            let optVal : string option = get.Optional.Field jsonName Decode.string
            box (optVal |> Option.map (fun s -> constructDU innerTi (box s)))
        | None ->
            if innerFn.Contains("String") then
                box (get.Optional.Field jsonName Decode.string)
            elif innerFn.Contains("Int32") then
                box (get.Optional.Field jsonName Decode.int)
            elif innerFn.Contains("Boolean") then
                box (get.Optional.Field jsonName Decode.bool)
            elif innerFn.Contains("FSharpList") then
                let innerInnerTi = if innerGs.Length > 0 then innerGs.[0] else innerTi
                let dec = Decode.list (decoderFor innerInnerTi)
                box (get.Optional.Field jsonName dec)
            else
                let dec : Decoder<obj> = decodeRecordObj innerTi
                box (get.Optional.Field jsonName dec)
    else
        let dec = decoderFor fieldTi
        get.Required.Field jsonName dec

// ============================================================
// Public API
// ============================================================

let inline encode (value: 'T) : JsonValue =
    encodeRecord typeof<'T> (box value)

let inline decode<'T> () : Decoder<'T> =
    decodeRecordObj typeof<'T> |> Decode.map unbox<'T>

File: apps/microblog/extension/fable_output/packages/hedge/src/Hedge/Codec.js
import { map as map_1, item } from "../../../../fable_modules/fable-library-js.4.29.0/Array.js";
import { nil, object, list as list_1 } from "../../../../fable_modules/Thoth.Json.10.2.0/Encode.fs.js";
import { ofArray, map } from "../../../../fable_modules/fable-library-js.4.29.0/List.js";
import { object as object_1, list as list_2, bool, int, string, map as map_2 } from "../../../../fable_modules/Thoth.Json.10.2.0/Decode.fs.js";
import { uncurry2 } from "../../../../fable_modules/fable-library-js.4.29.0/Util.js";
import { map as map_3 } from "../../../../fable_modules/fable-library-js.4.29.0/Option.js";

function camelCase(s) {
    if (s.length === 0) {
        return s;
    }
    else {
        return s[0].toLowerCase() + s.slice(1, s.length);
    }
}

function wrapperBaseType(fullname, generics) {
    if (fullname.indexOf("Interface.PrimaryKey") >= 0) {
        if ((generics.length > 0) && ((item(0, generics).fullname).indexOf("Int32") >= 0)) {
            return "int";
        }
        else {
            return "string";
        }
    }
    else if (fullname.indexOf("Interface.CreateTimestamp") >= 0) {
        return "int";
    }
    else if (fullname.indexOf("Interface.UpdateTimestamp") >= 0) {
        return "int";
    }
    else if (fullname.indexOf("Interface.SoftDelete") >= 0) {
        return "int";
    }
    else if (fullname.indexOf("Interface.ForeignKey") >= 0) {
        return "string";
    }
    else if (fullname.indexOf("Interface.RichContent") >= 0) {
        return "string";
    }
    else if (fullname.indexOf("Interface.Link") >= 0) {
        return "string";
    }
    else if (fullname.indexOf("Interface.Unique") >= 0) {
        if ((generics.length > 0) && ((item(0, generics).fullname).indexOf("Int32") >= 0)) {
            return "int";
        }
        else {
            return "string";
        }
    }
    else {
        return undefined;
    }
}

export function encodeValue(typeInfo, value) {
    const fn = typeInfo.fullname;
    const gs = typeInfo.generics || [];
    const matchValue = wrapperBaseType(fn, gs);
    if (matchValue == null) {
        if (fn.indexOf("String") >= 0) {
            return value;
        }
        else if (fn.indexOf("Int32") >= 0) {
            return value;
        }
        else if (fn.indexOf("Boolean") >= 0) {
            return value;
        }
        else if (fn.indexOf("FSharpList") >= 0) {
            const innerTi = (gs.length > 0) ? item(0, gs) : typeInfo;
            return list_1(map((value_6) => encodeValue(innerTi, value_6), value));
        }
        else {
            return encodeRecord(typeInfo, value);
        }
    }
    else if (matchValue === "int") {
        return value.fields[0];
    }
    else {
        return value.fields[0];
    }
}

export function encodeRecord(typeInfo, value) {
    return object(ofArray(map_1((tupledArg) => {
        const name = tupledArg[0];
        const fieldTi = tupledArg[1];
        const jsonName = camelCase(name);
        const fieldVal = value[name];
        const fn = fieldTi.fullname;
        const gs = fieldTi.generics || [];
        if ((fn.indexOf("FSharpOption") >= 0) && (gs.length > 0)) {
            const innerTi = item(0, gs);
            if ((fieldVal == null)) {
                return [jsonName, nil];
            }
            else {
                return [jsonName, encodeValue(innerTi, fieldVal)];
            }
        }
        else {
            return [jsonName, encodeValue(fieldTi, fieldVal)];
        }
    }, typeInfo.fields ? typeInfo.fields() : [])));
}

export function decoderFor(typeInfo) {
    const fn = typeInfo.fullname;
    const gs = typeInfo.generics || [];
    const matchValue = wrapperBaseType(fn, gs);
    if (matchValue == null) {
        if (fn.indexOf("String") >= 0) {
            return (path_4) => ((value_5) => map_2((value_4) => value_4, string, path_4, value_5));
        }
        else if (fn.indexOf("Int32") >= 0) {
            return (path_5) => ((value_7) => map_2((value_6) => value_6, uncurry2(int), path_5, value_7));
        }
        else if (fn.indexOf("Boolean") >= 0) {
            return (path_7) => ((value_10) => map_2((value_9) => value_9, bool, path_7, value_10));
        }
        else if (fn.indexOf("FSharpList") >= 0) {
            let d_5;
            const decoder = decoderFor((gs.length > 0) ? item(0, gs) : typeInfo);
            d_5 = ((path_8) => ((value_11) => list_2(uncurry2(decoder), path_8, value_11)));
            return (path_9) => ((value_13) => map_2((value_12) => value_12, uncurry2(d_5), path_9, value_13));
        }
        else {
            return decodeRecordObj(typeInfo);
        }
    }
    else if (matchValue === "int") {
        return (path) => ((value) => map_2((i) => (new (typeInfo.construct)(i)), uncurry2(int), path, value));
    }
    else {
        return (path_2) => ((value_2) => map_2((s) => (new (typeInfo.construct)(s)), string, path_2, value_2));
    }
}

export function decodeRecordObj(typeInfo) {
    return (path) => ((v) => object_1((get$) => {
        const values = map_1((tupledArg) => decodeFieldValue(get$, camelCase(tupledArg[0]), tupledArg[1]), typeInfo.fields ? typeInfo.fields() : []);
        return new (typeInfo.construct)(...values);
    }, path, v));
}

export function decodeFieldValue(get$, jsonName, fieldTi) {
    let objectArg_2, objectArg_3, objectArg_4, objectArg_5, objectArg_6, objectArg, objectArg_1;
    const fn = fieldTi.fullname;
    const gs = fieldTi.generics || [];
    if ((fn.indexOf("FSharpOption") >= 0) && (gs.length > 0)) {
        const innerTi = item(0, gs);
        const innerFn = innerTi.fullname;
        const innerGs = innerTi.generics || [];
        const matchValue = wrapperBaseType(innerFn, innerGs);
        if (matchValue == null) {
            if (innerFn.indexOf("String") >= 0) {
                return (objectArg_2 = get$.Optional, objectArg_2.Field(jsonName, string));
            }
            else if (innerFn.indexOf("Int32") >= 0) {
                return (objectArg_3 = get$.Optional, objectArg_3.Field(jsonName, uncurry2(int)));
            }
            else if (innerFn.indexOf("Boolean") >= 0) {
                return (objectArg_4 = get$.Optional, objectArg_4.Field(jsonName, bool));
            }
            else if (innerFn.indexOf("FSharpList") >= 0) {
                let dec;
                const decoder = decoderFor((innerGs.length > 0) ? item(0, innerGs) : innerTi);
                dec = ((path_3) => ((value_3) => list_2(uncurry2(decoder), path_3, value_3)));
                return (objectArg_5 = get$.Optional, objectArg_5.Field(jsonName, uncurry2(dec)));
            }
            else {
                const dec_1 = decodeRecordObj(innerTi);
                return (objectArg_6 = get$.Optional, objectArg_6.Field(jsonName, uncurry2(dec_1)));
            }
        }
        else if (matchValue === "int") {
            return map_3((i) => (new (innerTi.construct)(i)), (objectArg = get$.Optional, objectArg.Field(jsonName, uncurry2(int))));
        }
        else {
            return map_3((s) => (new (innerTi.construct)(s)), (objectArg_1 = get$.Optional, objectArg_1.Field(jsonName, string)));
        }
    }
    else {
        const dec_2 = decoderFor(fieldTi);
        const objectArg_7 = get$.Required;
        return objectArg_7.Field(jsonName, uncurry2(dec_2));
    }
}

